android/engine/src/main/java/org/smartregister/fhircore/engine/ui/questionnaire/QuestionnaireActivity.kt

```diff
@@ -138,6 +152,36 @@ open class QuestionnaireActivity : BaseMultiLanguageActivity(), View.OnClickList
     findViewById<Button>(R.id.btn_save_client_info).setOnClickListener(this)
   }
 
+  private fun setBarcode(questionnaire: Questionnaire, code: String, readonly: Boolean) {
+    questionnaire.find(QUESTIONNAIRE_ARG_BARCODE_KEY)?.apply {
+      initial =
+        mutableListOf(Questionnaire.QuestionnaireItemInitialComponent().setValue(StringType(code)))
+      readOnly = readonly
+    }
+  }
+
+  private fun Questionnaire.find(linkId: String): Questionnaire.QuestionnaireItemComponent? {
+    return item.find(linkId, null)
+  }
+  private fun List<Questionnaire.QuestionnaireItemComponent>.find(
+    linkId: String,
+    default: Questionnaire.QuestionnaireItemComponent?
+  ): Questionnaire.QuestionnaireItemComponent? {
+    var result = default
+    run loop@{
+      forEach {
+        if (it.linkId == linkId) {
+          result = it
+          return@loop
+        } else if (it.item.isNotEmpty()) {
+          result = it.item.find(linkId, result)
+        }
+      }
+    }
```

Comment:
I'm not a kotlin expert, but because you don't have any side effects before the bottom return, I think you could accomplish the same result in a less complex fashion with a non-local return in the conditional:

```suggestion
    forEach {
      if (it.linkId == linkId) {
        return it
      } else if (it.item.isNotEmpty()) {
        result = it.item.find(linkId, result)
      }
    }
```

Reply:
Yes, you're right that's better. Thanks